packages/block-editor/src/components/colors-gradients/control.js

```diff
@@ -0,0 +1,160 @@
+		if ( colorValue ) {
+			value = colorValue;
+			const colorObject = getColorObjectByColorValue( colors, value );
+			const colorName = colorObject && colorObject.name;
+			ariaLabel = sprintf( colorIndicatorAriaLabel, colorName || value );
+		}
+	} else if ( currentTab === 'gradient' && gradientValue ) {
+		value = gradientValue;
+		const gradientObject = __experimentalGetGradientObjectByGradientValue( gradients, value );
+		const gradientName = gradientObject && gradientObject.name;
+		ariaLabel = sprintf( gradientIndicatorAriaLabel, gradientName || value );
+	}
+
+	return (
+		<>
+			{ label }
+			{ !! value && (
+				<ColorIndicator
+					colorValue={ value }
+					aria-label={ ariaLabel }
+				/>
+			) }
+		</>
+	);
+}
+
+function ColorGradientControlInner( {
+	colors,
+	gradients,
+	disableCustomColors,
+	disableCustomGradients,
+	className,
+	label,
+	onColorChange,
+	onGradientChange,
+	colorValue,
+	gradientValue,
+} ) {
+	const canChooseAColor = onColorChange && ( ! isEmpty( colors ) || ! disableCustomColors );
+	const canChooseAGradient = onGradientChange && ( ! isEmpty( gradients ) || ! disableCustomGradients );
+	const [ currentTab, setCurrentTab ] = useState( gradientValue ? 'gradient' : !! canChooseAColor && 'color' );
+
+	if ( ! canChooseAColor && ! canChooseAGradient ) {
+		return null;
+	}
+	return (
+		<BaseControl
+			className={ classnames( 'block-editor-color-gradient-control', className ) }
+		>
+			<BaseControl.VisualLabel>
+				<VisualLabel
+					currentTab={ currentTab }
+					label={ label }
+					colorValue={ colorValue }
+					gradientValue={ gradientValue }
+				/>
+			</BaseControl.VisualLabel>
+			{ canChooseAColor && canChooseAGradient && (
+				<ButtonGroup className="block-editor-color-gradient-control__button-tabs">
+					<Button
+						isLarge
+						isPrimary={ currentTab === 'color' }
+						isSecondary={ currentTab !== 'color' }
+						onClick={ () => ( setCurrentTab( 'color' ) ) }
+					>
+						{ __( 'Solid Color' ) }
+					</Button>
+					<Button
+						isLarge
+						isPrimary={ currentTab === 'gradient' }
+						isSecondary={ currentTab !== 'gradient' }
+						onClick={ () => ( setCurrentTab( 'gradient' ) ) }
+					>
+						{ __( 'Gradient' ) }
+					</Button>
+				</ButtonGroup>
+			) }
+			{ currentTab === 'color' && (
+				<ColorPalette
+					value={ colorValue }
+					onChange={ canChooseAGradient ?
+						( newColor ) => {
+							onColorChange( newColor );
+							onGradientChange();
+						} :
+						onColorChange
+					}
+					{ ... { colors, disableCustomColors } }
+				/>
+			) }
+			{ currentTab === 'gradient' && (
+				<GradientPicker
+					value={ gradientValue }
+					onChange={ canChooseAGradient ?
+						( newGradient ) => {
+							onGradientChange( newGradient );
+							onColorChange();
+						} :
+						onColorChange
+					}
+					{ ... { gradients, disableCustomGradients } }
+				/>
+			) }
+		</BaseControl>
+	);
+}
+
+function ColorGradientControlSelect( props ) {
```

Comment:
Observation: some of these component names aren't super clear. It's not a big deal, since this is all private to this file. Some of these are also tied to implementation details (e.g. one of the components reads from `useSelect`).

Maybe something to keep in mind and polish later if the problem domain is simplifiable.

Reply:
> Observation: some of these component names aren't super clear. It's not a big deal, since this is all private to this file. Some of these are also tied to implementation details (e.g. one of the components reads from useSelect).

> Maybe something to keep in mind and polish later if the problem domain is simplifiable.

That's true; the names are not very clean. These components are tied to implementation details (e.g., useSelect) because they exist to overcome a hooks limitation: hooks should un-unconditionally be called in the same order. In this case, in some situations, we don't need to call useSelect because the props are already available. To overcome the hooks limitation, we have a component that uses useSelect and other that does not, depending on the situation, we use a component or the other. I will try to research/test better solutions to this problem after this PR is merged as I think It will become common soon.